src/expr.rs

```diff
@@ -722,6 +742,20 @@ fn simplify_mul(es: &[Expr]) -> Expr {
         result.push(Expr::F64(lit));
     }
     result.sort_by(|a, b| a.partial_cmp(b).unwrap());
+    let mut i = 0;
```

Comment:
Nice, that gets rid of some annoying terms, but might benefit from some generalisation.
1. add some tests that check cases like `a * b * a^-1` where the cancelling powers are separated. I assume this might be missed currently
2. also misses `1/a * a = 1`
3. consider sorting by the base and grouping equal bases s.t. `a^n b^m a^k` -> `a^n a^k b^m`; then you can collect powers and simplify exponents, i.e `a^(n + k) b^m`. Note that you might be forced to temporarily 'up-convert' `x` to `x^1`. This should give you the specific optimisation, plus many more like `a^-2 b a a = b`
4. more tests ;)

Reply:
I think the `1/a * a = 1` case is handled because the multiplication args are always sorted iirc. But that should be a test anyway

Reply:
This is in main as of now.